src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java

```diff
@@ -2907,19 +2917,86 @@ private SocketFactory getSocketFactory() throws IOException {
         return socketFactory;
     }
 
-    // This method contains the old logic of connecting to
-    // a socket of one of the IPs corresponding to a given host name.
-    // In the old code below, the logic around 0 timeout has been removed as
-    // 0 timeout is not allowed. The code has been re-factored so that the logic
-    // is common for hostName or InetAddress.
-    private Socket getDefaultSocket(String hostName, int portNumber, int timeoutInMilliSeconds) throws IOException {
-        // Open the socket, with or without a timeout, throwing an UnknownHostException
-        // if there is a failure to resolve the host name to an InetSocketAddress.
-        //
+    /**
+     * Helper function which traverses through queue of InetAddresses to find a resolved one
+     * 
+     * @param addrq
+     *        Address queue
+     * @param portNumber
+     *        Port Number
+     * @return First resolved queue or unresolved queue if none found
+     * @throws IOException
+     */
+    private InetSocketAddress traverseAddressQueue(Queue<InetAddress> addrq, int portNumber) throws IOException {
+        InetSocketAddress addr = new InetSocketAddress(portNumber);
+        while (addrq.peek() != null) {
+            addr = new InetSocketAddress(addrq.poll(), portNumber);
+            if (!addr.isUnresolved())
+                return addr;
+        }
+        return addr;
+    }
+
+    /**
+     * This method makes a connected socket given a hostname or IP address.
+     * 
+     * @param hostName
+     *        Hostname or IP Address
+     * @param portNumber
+     *        Port number
+     * @param timeoutInMilliSeconds
+     *        Timeout
+     * @param IPAddressPreference
+     *        IPv4First (default), IPv6First or UsePlatformDefault
+     * @return Connected Socket
+     * @throws IOException
+     * @throws SQLServerException
+     */
+    private Socket getSocketByIPvPreference(String hostName, int portNumber, int timeoutInMilliSeconds,
+            String IPAddressPreference) throws IOException, SQLServerException {
+        InetSocketAddress addr = null;
+        InetAddress addresses[] = InetAddress.getAllByName(hostName);
+        Queue<InetAddress> addrq = null;
+
+        IPvAddressPreferenceEnum IPvPref = IPvAddressPreferenceEnum.valueOfString(IPAddressPreference);
+
+        switch (IPvPref) {
+            case IPv6First:
+                // Try to connect to IPv6 Addresses
+                addrq = getIPv6AddressQueue(addresses);
+                addr = traverseAddressQueue(addrq, portNumber);
+                if (addr.isUnresolved())
+                    return getConnectedSocket(addr, timeoutInMilliSeconds);
+                // No IPv6 Addresses can be connected to, try IPv4 addresses now
+                addrq = getIPv4AddressQueue(addresses);
```

Comment:
prev comment was not addressed - 
these are all doing the same thing, why do we need to get q, traverse q all over the place? can be refactored to just a getIpAddress which checks preference option and gets the addr from the whatever q.

Reply:
We could refactor all the functionality to exist in an new overload of getConnectedSocket, but I think decomposing the logic into smaller functions is better than having one big 100 line function.

Reply:
Changed as recommended by Terry